relayer/canonical_validator_client.go

```diff
@@ -34,10 +34,51 @@ func (v *CanonicalValidatorClient) GetSubnetID(ctx context.Context, chainID ids.
 	return v.client.ValidatedBy(ctx, chainID)
 }
 
+// Gets the current validator set of the given subnet ID, include the validators' BLS public keys.
+// This implementation of GetValidatorSet currently makes two RPC requests, one to get the
+// subnet validators, and another to get their BLS public keys. This is necessary in order to enable
+// the use of the public APIs (which don't support "GetValidatorsAt") because BLS keys are currently
+// only associated with primary network validation periods. If ACP-13 is implementated in the future
+// (https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/13-subnet-only-validators.md), it may
+// become possible to reduce this to a single RPC request that returns both the subnet validators
+// as well as their BLS public keys.
 func (v *CanonicalValidatorClient) GetValidatorSet(
 	ctx context.Context,
 	height uint64,
 	subnetID ids.ID,
 ) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
-	return v.client.GetValidatorsAt(ctx, subnetID, height)
+	// Get the current subnet validators. These validators are not expected to include
+	// BLS signing information given that addPermissionlessValidatorTx is only used to
+	// add primary network validators.
+	subnetVdrs, err := v.client.GetCurrentValidators(ctx, subnetID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look up the primary network validators of the NodeIDs validating the subnet
+	// in order to get their BLS keys.
+	res := make(map[ids.NodeID]*validators.GetValidatorOutput, len(subnetVdrs))
+	subnetNodeIDs := make([]ids.NodeID, 0, len(subnetVdrs))
+	for _, subnetVdr := range subnetVdrs {
+		subnetNodeIDs = append(subnetNodeIDs, subnetVdr.NodeID)
+		res[subnetVdr.NodeID] = &validators.GetValidatorOutput{
+			NodeID: subnetVdr.NodeID,
+			Weight: subnetVdr.Weight,
+		}
+	}
+	primaryVdrs, err := v.client.GetCurrentValidators(ctx, ids.Empty, subnetNodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set the BLS keys of the result.
+	for _, primaryVdr := range primaryVdrs {
+		vdr, ok := res[primaryVdr.NodeID]
+		if !ok {
```

Comment:
can we add documentation that if we have a validator taht we don't find it's public key in primary validator set we still include it for mapping, so it'd affect total stake weight, and need to check whether it has a public key

Reply:
Yep, I'll add it below. Worth noting that this case here is actually just if the "GetCurrentValidators" returns an unexpected node ID that wasn't included in the filter in the request body. I'll document that also.